Aplib.Core/Belief/Belief.cs

```diff
@@ -11,9 +11,12 @@ namespace Aplib.Core.Belief
     /// It implements the <see cref="IBelief"/> interface.
     /// It supports implicit conversion to <typeparamref name="TObservation"/>.
     /// </remarks>
-    /// <typeparam name="TReference">The type of the object reference used to generate/update the observation.</typeparam>
+    /// <typeparam name="TReference">
+    /// The type of the object reference used to generate/update the observation. This <i>must</i> be a reference type,
+    /// be aware that this is not enforced by C# if <typeparamref name="TReference"/> is an interface.
+    /// </typeparam>
     /// <typeparam name="TObservation">The type of the observation that the belief represents.</typeparam>
-    public class Belief<TReference, TObservation> : IBelief
+    public class Belief<TReference, TObservation> : IBelief where TReference : class
```

Comment:
We have discussed this extensively on Discord, but if we don't allow structs disguised as interfaces, we should definitely throw an argument exception if we detect the type is a value type by using the `IsValueType` boolean. https://learn.microsoft.com/en-us/dotnet/api/system.type.isvaluetype?view=netstandard-2.1